### DISCUSSION TITLE
Upload to specific folder example

### DISCUSSION THREAD

**Discussion by 0-don** [qgustavor/mega#105] [Ideas]
I would like to know how to upload a file to specific folder?

```ts
  const storage = await new Storage({
    email: process.env.MEGA_EMAIL,
    password: process.env.MEGA_PASSWORD,
  }).ready;

  let folder = storage.root.children?.find(
    (e) => e.name === process.env.MEGA_DIR
  );

  if (!folder) folder = await storage.mkdir({ name: process.env.MEGA_DIR });

  const file = await storage.upload(fileName, jsonFile)
    .complete;

  await file.moveTo(folder);

  const link = await file.link({ noKey: false });
  return link;
```

currently I am doing it like this but I wanted to ask if there is an easier way?

it would be nice to do something like this so it parses the string for the filename and creates the missing folders if they dont exist.
```js
const file = await storage.upload("./folder/folder1/" + fileName, jsonFile)
```

**Comment by qgustavor** [comment]
You don't need to call `storage.upload` then `file.moveTo(folder)`, just call `folder.upload`.

About the upload shorthand and in general, at the moment the library doesn't have any path transversal functions since it's not straightforward since the library lets you chose any name for files, even files with invalid characters under most file systems such as `\` and `/`. What to do in those cases?

I will leave this discussion in the Ideas folder. If anyone wants to implement path transversal functions and integrate those in the library is free to send a pull request.

**Reply by 0-don** [reply]
how would it work if its a nested directory and you have to create the folders first?

**Reply by 0-don** [reply]
![image](https://user-images.githubusercontent.com/70039285/175648068-8f0a3866-3141-4b03-a77f-770212816ca7.png)
folder.upload returns a Writeble type and not a promise so I cant create a link afterwards

**Reply by qgustavor** [reply]
About the return type: check the example [from the home page](https://mega.js.org/), just use the `.complete` property.

```javascript
const file = await storage.upload("./folder/folder1/" + fileName, jsonFile).complete 
// ^^^ will not work with current code, is just a proposal ^^^
const link = await file.link()
```

About adding a shortcut for the `.link()` directly from the `.upload()`: I'm against it, at it's currently implemented it just adds one extra line to the code.

About handling nested paths: yeah, the proposal is that there would be an helper function would parse the file name and create the folders automatically, and the upload (among others that receive file names as input) would call this function before uploading. With the current code you need to implement this yourself.

**Reply by FouadRaheb** [reply]
Shouldn't the return type of `mutableFile.upload()` be `UploadStream` instead of `Writable`? Since `.complete` is only available on the subclass `UploadStream`.

I'm currently just casting the return type `const file = await (dir.upload(opts) as UploadStream).complete;`

Also `allowUploadBuffering` does not exist on `uploadOpts`

**Reply by qgustavor** [reply]
I rarely use Typescript, so I guess the typings might be wrong. You can send a PR. Well, in fact ES module typings might be even more wrong since they assume a Deno environment.
